apis/observability/v1alpha1/telemetry_types.go

```diff
@@ -0,0 +1,174 @@
+
+// +kubebuilder:object:root=true
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Selected",type="string",JSONPath=`.metadata.annotations.sharedtelemetryconfig\.internal\.spaces\.upbound\.io/selected`
+// +kubebuilder:printcolumn:name="Failed",type="string",JSONPath=`.metadata.annotations.sharedtelemetryconfig\.internal\.spaces\.upbound\.io/failed`
+// +kubebuilder:printcolumn:name="Provisioned",type="string",JSONPath=`.metadata.annotations.sharedtelemetryconfig\.internal\.spaces\.upbound\.io/provisioned`
+// +kubebuilder:printcolumn:name="Phase",type="string",JSONPath=".status.phase"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:scope=Namespaced,categories=observability,shortName=stc
+
+// SharedTelemetryConfig defines a telemetry configuration over a set of ControlPlanes.
+type SharedTelemetryConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:XValidation:rule="!has(self.pipeline) || !has(self.pipeline.metrics) || self.pipeline.metrics.all(x, self.exporters.exists(p, p == x))",message="spec.pipeline.metrics values must be present in spec.exporters"
+	// +kubebuilder:validation:XValidation:rule="!has(self.pipeline) || !has(self.pipeline.traces) || self.pipeline.traces.all(x, self.exporters.exists(p, p == x))",message="spec.pipeline.traces values must be present in spec.exporters"
+	Spec   SharedTelemetryConfigSpec   `json:"spec"`
+	Status SharedTelemetryConfigStatus `json:"status,omitempty"`
+}
+
+// SharedTelemetryConfigList contains a list of SharedTelemetryConfigs.
+//
+// +kubebuilder:object:root=true
+type SharedTelemetryConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []SharedTelemetryConfig `json:"items"`
+}
+
+// SharedTelemetryConfigSpec defines a telemetry configuration over a set of ControlPlanes.
+type SharedTelemetryConfigSpec struct {
+	// ControlPlaneSelector defines the selector for ControlPlanes on which to
+	// configure telemetry.
+	// +kubebuilder:validation:XValidation:rule="(has(self.labelSelectors) || has(self.names)) && (size(self.labelSelectors) > 0 || size(self.names) > 0)",message="either names or a labelSelector must be specified"
+	ControlPlaneSelector spacesv1alpha1.ResourceSelector `json:"controlPlaneSelector"`
+
+	// Exporters defines the exporters to configure on the selected ControlPlanes.
+	// Untyped as we use the underlying OpenTelemetryOperator to configure the
+	// OpenTelemetry collector's exporters. Use the OpenTelemetry Collector
+	// documentation to configure the exporters.
+	// Currently only supported exporters are push based exporters.
+	// +kubebuilder:validation:MaxProperties=20
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Exporters map[string]Export `json:"exporters"`
```

Comment:
Why not the following and remove `export.go` ? 

```suggestion
	Exporters map[string]common.JSONObject `json:"exporters"`
```

Reply:
no need for the alias, you can set +kubebuilder:pruning:PreserveUnknownFields as you are already doing here and that will work.

Reply:
I thought so as well but it doesnt. If I generate it with `map[string]common.JSONObject`, the `x-kubernetes-preserve-unknown-fields: true` gets removed from the `additionalProperties`. And then nested objects under `exporters` dont work.

```
                   additionalProperties:
-                    description: Export defines the telemetry exporter configuration.
+                    description: |-
+                      JSONObject represents any valid JSON value of an object.
+                      These types are supported: bool, int64, float64, string, []interface{}, map[string]interface{} and nil.
                     type: object
-                    x-kubernetes-preserve-unknown-fields: true
```
For example this works:
```yaml
  exporters:
    otlphttp: {}
```

But this fails:
```yaml
  exporters:
    otlphttp:
      headers:
         api-key: bob
```
with 
```
": strict decoding error: unknown field "spec.exporters.otlphttp.endpoint", unknown field "spec.exporters.otlphttp.headers"
```

Reply:
Why map[string]JSONObject and not just JSONObject? 🤔

Reply:
The CEL rule breaks then because its not a map.

Reply:
I am working on adding admission webhooks for more advanced validations (i.e. [validating](https://github.com/upbound/up-sdk-go/pull/76#discussion_r1572899359) the export configurations), we can drop CEL rules and cover this validation there in code. 

So, we can simplify what is here as in the first comment in that case?

Reply:
Great. Yeah we can simplify it then 🎉

Reply:
Pushed a commit.